lib/membrane/rtp/tcp_depayloader.ex

```diff
@@ -0,0 +1,119 @@
+defmodule Membrane.RTP.TCP.Depayloader do
+  @moduledoc """
+  This element provides functionality of depayloading RTP Packets received by TCP and redirecting
+  RTSP messages received in the same stream. The encapsulation is described in RFC 7826 Section 14.
+  """
+  use Membrane.Filter
+
+  alias Membrane.{Buffer, RemoteStream, RTP, RTSP}
+
+  def_options rtp_channel_id: [
+                spec: non_neg_integer(),
+                default: 0,
+                description: """
+                Channel identifier which encapsulated RTP packets will have.
+                """
+              ],
+              rtsp_session: [
+                spec: pid() | nil,
+                default: nil,
+                description: """
+                PID of a RTSP Session (returned from Membrane.RTSP.start or Membrane.RTSP.start_link)
+                that received RTSP responses will be forwarded to. If nil the responses will be
+                discarded.
+                """
+              ]
+
+  def_input_pad :input, accepted_format: %RemoteStream{type: :bytestream}
+
+  def_output_pad :output, accepted_format: %RemoteStream{type: :packetized, content_format: RTP}
+
+  @impl true
+  def handle_init(_ctx, opts) do
+    state =
+      Map.from_struct(opts)
+      |> Map.merge(%{
+        unprocessed_data: <<>>
+      })
+
+    {[], state}
+  end
+
+  @impl true
+  def handle_stream_format(:input, _stream_format, _ctx, state) do
+    stream_format = %RemoteStream{type: :packetized, content_format: RTP}
+    {[stream_format: {:output, stream_format}], state}
+  end
+
+  @impl true
+  def handle_buffer(
+        :input,
+        %Buffer{payload: payload, metadata: metadata},
+        _ctx,
+        %{unprocessed_data: unprocessed_data, rtsp_session: rtsp_session} = state
+      ) do
+    unprocessed_data =
+      if rtsp_response?(unprocessed_data, payload) do
+        if rtsp_session != nil do
+          {:ok, %RTSP.Response{status: 200}} =
+            RTSP.handle_response(rtsp_session, unprocessed_data)
+        end
+
+        <<>>
+      else
+        unprocessed_data
+      end
+
+    packets_binary = unprocessed_data <> payload
+
+    {unprocessed_data, complete_packets_binaries} =
+      get_complete_packets(packets_binary, state.rtp_channel_id)
+
+    packets_buffers =
+      Enum.map(complete_packets_binaries, &%Buffer{payload: &1, metadata: metadata})
```

Comment:
do we have a guarantee that the `metadata` is rewritten properly? What if the buffer's payload is put in `unprocessed_data`, and the metadata is attached to the data corresponding to the previous input buffer?

Reply:
the only metadata that is in buffers coming from TCP Source are `tcp_source_address`, `tcp_source_port` and `arrival_ts`. The first two don't change and the third one can be interpreted as in when the packet has fully arrived, which then works in this case

Reply:
I see, it makes sense, however, I am slightly worried that in case we were about to extend the `metadata` sent from the `TCP Source`, we would need to remember that the metadata is translated by one buffer when passed through the depayloader